ykcompile/src/lib.rs

```diff
@@ -596,25 +758,30 @@ impl<TT> TraceCompiler<TT> {
         op1: &Operand,
         op2: &Operand,
     ) -> Result<(), CompileError> {
-        // Move `op1` into `dest`.
+        // The value of the addition is stored in the first field of the result tuple.
+        let mut val_dest = dest.clone();
+        val_dest.projection.push(Projection::Field(0));
+
+        // Move `op1` into `val_dest`.
         match op1 {
-            Operand::Place(p) => self.mov_place_place(dest, &p)?,
-            Operand::Constant(Constant::Int(ci)) => self.mov_place_constint(dest, &ci)?,
+            Operand::Place(p) => self.mov_place_place(&val_dest, &p)?,
+            Operand::Constant(Constant::Int(ci)) => self.mov_place_constint(&val_dest, &ci)?,
             Operand::Constant(Constant::Bool(_b)) => unreachable!(),
             Operand::Constant(c) => return Err(CompileError::Unimplemented(format!("{}", c))),
         };
-        // Add together `dest` and `op2`.
-        let lloc = self.place_to_location(dest)?;
+        // Add together `val_dest` and `op2`.
+        let lloc = self.place_to_location(&val_dest)?;
```

Comment:
I think this means that we currently only allocate the first field of the tuple (and never the tuple itself). Which at the moment is fine, since those are equivalent. Though I'm not sure if this could cause problems in the future.

I think ideally at some point the tuple should be allocated somewhere, and `val_dest` should just be the location of the tuple + the offset of the field. Which makes me think that we might need a `Location::Mem(mem, offset)` in the future.

Reply:
Good spot! I'd like to fix this now. It will surely bite use later!

Reply:
But hold on. 

```
let lloc = self.place_to_location(&val_dest)?;
```

^ still results in a call to `local_to_location` which should allocate the whole tuple.

Do you agree?

Reply:
Yes, I wasn't entirely right. `local_to_location` ignores the projections so it will allocate the `Local` of the tuple. What we need to do later though is, if the projection is bigger than `0`, we need to update the offset. So let's say the tuple is allocated in `Location::Stack(offset)`, then we need to return `Location::Stack(offset + fieldidx * size)`.

What happens if the tuple was squeezed into a register though? Can we use `al, ax, eax` to access the fields, or do we need to move the tuple to the stack first?

Reply:
> What we need to do later though is, if the projection is bigger than 0, we need to update the offset.

place_to_location almost handles that already with:
```
Projection::Field(idx) => Ok(Location::new_mem(RDI.code(), (idx * 8) as i32))
```

`idx * 8` is wrong. I'll fix that. I also think this should panic if there is more than one projection in the chain.

> What happens if the tuple was squeezed into a register though? 

That can't happen for now. `can_live_in_register()` returns false for tuples.

Reply:
> place_to_location almost handles that

It's similar but not the same. We won't have `RDI` here, but `local_to_location` will return a `Location` to which we need to add the offset of the field.

Reply:
I think that's the fix I'm writing. Let me finish it and we can see.

Reply:
Give that last push a shot.